library/std/src/sys/unix/process/process_unix.rs

```diff
@@ -107,6 +97,95 @@ impl Command {
         }
     }
 
+    // Attempts to fork the process. If successful, returns
+    // Ok((0, -1)) in the child, and Ok((child_pid, child_pidfd)) in the parent.
+    fn do_fork(&mut self) -> Result<(libc::c_long, libc::pid_t), io::Error> {
+        // Whatever happens after the fork is almost for sure going to touch or
+        // look at thbe environment in one way or another (PATH in `execvp` or
+        // accessing the `environ` pointer ourselves). Make sure no other thread
+        // is accessing the environment when we do the fork itself.
+        //
+        // Note that as soon as we're done with the fork there's no need to hold
+        // a lock any more because the parent won't do anything and the child is
+        // in its own process.
+        let _env_lock = unsafe { sys::os::env_lock() };
+
+        // If we fail to create a pidfd for any reason, this will
+        // stay as -1, which indicates an error
+        let mut pidfd: libc::pid_t = -1;
+
+        // On Linux, attempt to use the `clone3` syscall, which
+        // supports more argument (in prarticular, the ability to create a pidfd).
+        // If this fails, we will fall through this block to a call to `fork()`
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                static HAS_CLONE3: AtomicBool = AtomicBool::new(true);
+
+                const CLONE_PIDFD: u64 = 0x00001000;
+
+                #[repr(C)]
+                struct clone_args {
+                    flags: u64,
+                    pidfd: u64,
+                    child_tid: u64,
+                    parent_tid: u64,
+                    exit_signal: u64,
+                    stack: u64,
+                    stack_size: u64,
+                    tls: u64,
+                    set_tid: u64,
+                    set_tid_size: u64,
+                    cgroup: u64,
+                }
+
+                syscall! {
+                    fn clone3(cl_args: *mut clone_args, len: libc::size_t) -> libc::c_long
+                }
+
+                if HAS_CLONE3.load(Ordering::Relaxed) {
+                    let mut flags = 0;
+                    if self.make_pidfd {
+                        flags |= CLONE_PIDFD;
+                    }
+
+                    let mut args = clone_args {
+                        flags,
+                        pidfd: &mut pidfd as *mut libc::pid_t as u64,
+                        child_tid: 0,
+                        parent_tid: 0,
+                        exit_signal: libc::SIGCHLD as u64,
+                        stack: 0,
+                        stack_size: 0,
+                        tls: 0,
+                        set_tid: 0,
+                        set_tid_size: 0,
+                        cgroup: 0
+                    };
+
+                    let args_ptr = &mut args as *mut clone_args;
+                    let args_size = crate::mem::size_of::<clone_args>();
+
+                    let res = cvt(unsafe { clone3(args_ptr, args_size) });
```

Comment:
This won't run `atfork` callbacks. This is for example used by `rand` to force all `ReseedingRng` instances to reseed themselves. (https://github.com/rust-random/rand/blob/42247d8616b0f8fb62d8a5a6e8dd8f6a17e1b3eb/src/rngs/adapter/reseeding.rs#L287) The `rand` usage is not relevant here, but there may be other cases.

Reply:
I had thought that `posix_spawnnp` didn't run `atfork` handlers, but it looks like I was mistaken.
This will require some further thought.

Reply:
The reason that I used `__clone3` is that glibc's `clone()` implementation requires the caller to allocate a stack for the new process (even though the kernel allows passing a null stack pointer to the raw `clone` syscall). We could allocate a stack ourself, but I think it would be better to avoid that if at all possible.

Reply:
Actually, the `posix_spawn` man page states:

```
       According to POSIX, it is unspecified whether fork handlers
       established with pthread_atfork(3) are called when posix_spawn() is
       invoked.  Since glibc 2.24, the fork handlers are not executed in any
       case.  On older implementations, fork handlers are called only if the
       child is created using fork(2).
```

so I think this should be fine

Reply:
There is some extensive discussion going on in glibc upstream about handling and wrapping clone3. See:
- https://www.linuxplumbersconf.org/event/7/contributions/840/attachments/691/1277/The_Clone_Wars.pdf
- Brief discussion at https://www.youtube.com/watch?v=VQ89O7Z5prc#t=2h59m
- Full talk at https://www.youtube.com/watch?v=VQ89O7Z5prc#t=3h25m30s